uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java

```diff
@@ -0,0 +1,1144 @@
+        }
+
+        @Override
+        public void onAgentDeattached(AgentControl agentControl) {
+            log.info("Agent {} is disconnected", agentControl.getAgentId());
+        }
+    };
+
+    private final AgentControl.ConnectionEvents connectionEvents = new AgentControl.ConnectionEvents() {
+        @Override
+        public void onMessageReceived(ConnectionControl connectionControl, Mqtt5Message message) {
+            eventStorage.addEvent(new MqttMessageEvent(connectionControl, message));
+            log.info("Message received on connection with name {}: {}", connectionControl.getConnectionName(), message);
+        }
+
+        @Override
+        public void onMqttDisconnect(ConnectionControl connectionControl, Mqtt5Disconnect disconnect, String error) {
+            // TODO: also add to eventStore
+            log.info("MQTT client disconnected. Error: {}", error);
+        }
+    };
+
+    /**
+     * Creates instance of MqttControlSteps.
+     *
+     * @param testContext the instance of TestContext
+     * @param scenarioContext the instance of ScenarioContext
+     * @param registrationContext the instance of RegistrationContext
+     * @param resources the instance of AWSResources
+     * @param resourcesContext the instance of AWSResourcesContext
+     * @param iotSteps the instance of IotSteps
+     * @param engineControl the MQTT clients control
+     * @param eventStorage the MQTT event storage
+     * @param greengrassClient the GreengrassV2Client instance
+     * @throws IOException on IO errors
+     */
+    @Inject
+    public MqttControlSteps(
+            TestContext testContext,
+            ScenarioContext scenarioContext,
+            RegistrationContext registrationContext,
+            AWSResources resources,
+            AWSResourcesContext resourcesContext,
+            IotSteps iotSteps,
+            EngineControl engineControl,
+            EventStorageImpl eventStorage,
+            GreengrassV2Client greengrassClient) throws IOException {
+        this.testContext = testContext;
+        this.scenarioContext = scenarioContext;
+        this.registrationContext = registrationContext;
+        this.resources = resources;
+        this.resourcesContext = resourcesContext;
+        this.iotSteps = iotSteps;
+        this.engineControl = engineControl;
+        this.eventStorage = eventStorage;
+        this.greengrassClient = greengrassClient;
+
+        initMqttVersions();
+        startMqttControl();
+    }
+
+    /**
+     * Associate client device with a core device .
+     *
+     * @param clientDeviceId string user defined client device id
+     */
+    @When("I associate {string} with ggc")
+    public void associateClient(String clientDeviceId) {
+        final String coreName = testContext.coreThingName();
+        final String clientDeviceThingName = getClientDeviceThingName(clientDeviceId);
+        greengrassClient.batchAssociateClientDeviceWithCoreDevice(b -> b.coreDeviceThingName(coreName)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code makes a call to a batch operation, but it does not handle any errors returned by that batch operation. Batch operations can succeed without throwing exceptions even if some items in the request fail. Consider checking the return value of the batch operation for failures.